File: LeetCode/Apple/TopQuestions/Recursion/subsets.py
class Solution:
    def subsets(self,nums):
        #edge case
        if nums is None or len(nums)==0:
            return None
        res = [[]]
        for num in nums:
            n=len(res)
            for i in range(n):
                res.append(res[i]+[num])
        return res

#Time Complexity: O(len(num)*)
#Space Complesity: O(1)O()

if __name__ == "__main__":
    nums = [1,2,3]
    output = Solution()
    print ("The subsets are:",output.subsets(nums))

File: LeetCode/Apple/TopQuestions/Recursion/subsets.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;
class Solution
{
public:
    vector<vector<int>> subsets(vector<int>& nums)
    {
      vector<vector<int>> v;
        vector<int> b;
        v.push_back(b);
        for(int i=0;i<nums.size();i++)
        {
          int t=v.size();
            for(int j=0;j<t;j++)
            {
                b=v[j];
                b.push_back(nums[i]);
                v.push_back(b);
            }
        }
        return v;
    }
};
int main()
{
  vector <int> nums = {1,2,3};
  vector <vector <int>> output;
  Solution S1;
  output=S1.subsets(nums);
  //print the result to see the outcome
  return 0;
}
